src/main/java/dev/dbos/transact/database/DatabaseUrl.java

```diff
@@ -0,0 +1,122 @@
+package dev.dbos.transact.database;
+
+import java.net.URI;
+import java.net.URLDecoder;
+import java.net.URLEncoder;
+import java.nio.charset.StandardCharsets;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+public record DatabaseUrl(
+    String scheme,
+    String host,
+    int port,
+    String database,
+    String username,
+    String password,
+    Map<String, String> query) {
+
+  public DatabaseUrl {
+    Objects.requireNonNull(scheme);
+    Objects.requireNonNull(host);
+    Objects.requireNonNull(database);
+  }
+
+  public DatabaseUrl withDatabase(String database) {
+    return new DatabaseUrl(scheme(), host(), port(), database, username(), password(), query());
+  }
+
+  public DatabaseUrl withUser(String user) {
+    return new DatabaseUrl(scheme(), host(), port(), database(), user, password(), query());
+  }
+
+  public DatabaseUrl withPassword(String password) {
+    return new DatabaseUrl(scheme(), host(), port(), database(), username(), password, query());
+  }
+
+  static Map<String, String> parseQuery(String query) {
```

Comment:
Is there a standard library in Java to parse URLs? This looks fine but I'm worried about weird strings in the wild, like the one with url encoded that may have random `&` in it.

Reply:
Apparently. the PG JDBC driver has a parseURL function. We already have a dependency on that so I'll see if we can reuse it.

Reply:
https://github.com/pgjdbc/pgjdbc/blob/ef6692365886ad772d1810da568e1ed147906aa4/pgjdbc/src/main/java/org/postgresql/Driver.java#L535

Reply:
parseURL in the PG JDBC driver is only for parsing JDBC PG URLs